### DISCUSSION TITLE
Context API error in Next.js

### DISCUSSION THREAD

**Discussion by kaitok** [vercel/next.js#51280] [Help] [answered]
### Summary

I am working with a self-made UI library called mono-ui in Next.js, and I am using ContextAPI to implement dark mode, but when I call Provider in Next.js, I get the following error.
code: [mono-ui](https://github.com/kaitok/mono-ui)

```
Unhandled Runtime Error
TypeError: null is not an object (evaluating 'dispatcher.useState')
```
<img width="774" alt="Screenshot 2023-06-14 at 18 01 46" src="https://github.com/vercel/next.js/assets/5301304/117e6774-135c-47dd-847f-65244f86862b">

I have mono-ui configured locally with npm link and have tried rendering simple divs with success. But when I use useState in ThemeProvider.tsx, I get an error.

By the way, I placed the same code in the Next.js project under the name ThemeProviderLocal and it worked fine.
[Providers.tsx](https://github.com/kaitok/mono-ui/blob/main/demo/src/app/Providers.tsx)
```jsx
'use client'
import React from 'react'
import { ThemeProvider } from '@kaitok/mono-ui'
// import { ThemeProvider } from './ThemeProviderLocal'

export function Providers({ children }: { children: React.ReactNode }) {
  return <ThemeProvider>{children}</ThemeProvider>
}
```

Provider in UI library
[ThemeProvider.tsx](https://github.com/kaitok/mono-ui/blob/main/src/ThemeProvider.tsx)
```jsx
import React, { createContext, useState } from 'react'

const ThemeContext = createContext({
  theme: 'light',
  toggleTheme: () => {},
})

const ThemeProvider = ({ children }: { children: React.ReactNode }) => {
  const initTheme = 'light'
  const [theme, setTheme] = useState(initTheme)
  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'))
  }

  const contextValue = {
    theme,
    toggleTheme,
  }

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  )
}

export { ThemeProvider, ThemeContext }
```

**Comment by icyJoseph** [comment]
The ThemeProvider must also be within a 'use client' tree, either by its importer parent, or explicitly set by itself.

If you import ThemeProvider, as is, into a Server Component, React won't have the client side hooks. The same applies, for any potential `useTheme` etc.

**Reply by kaitok** [reply]
Thank you!
I understand that ThemeProvider.tx is imported in Providers.tx and is called with 'use client', but is there a different way to set this up?

**Reply by icyJoseph** [reply]
Aha right I saw it in reverse... my bad. Where do  you use ThemeContext though? I reckon it ends up as part of a server component tree.

**Reply by kaitok** [reply]
ThemeContext is used in page.tsx
https://github.com/kaitok/mono-ui/blob/main/demo/src/app/page.tsx

```jsx

'use client'
import React, { useContext } from 'react'
import { Button } from '@kaitok/mono-ui'
// import { ThemeContext } from './ThemeProviderLocal'
import { ThemeContext } from '@kaitok/mono-ui'

export default function Home() {
  const { theme, toggleTheme } = useContext(ThemeContext)
  const handleClick = () => {
    toggleTheme()
  }
  return (
    <div>
      <Button label={'button'} width={'200px'} />
      theme: {theme}
      <br />
      <button onClick={handleClick}>Toggle Theme</button>
      <br />
    </div>
  )
}
```

**Reply by icyJoseph** [reply]
I always get this things upside down, but isn't your library packaging all of React with it? I think you need to make it external. Otherwise, context's can't be shared.

I got it to work, but now I am missing the Button component 😅 stuff like this:

```console
./src/app/page.tsx
Attempted import error: 'ThemeContext' is not exported from '@kaitok/mono-ui' (imported as 'ThemeContext').

Import trace for requested module:
./src/app/page.tsx

./src/app/page.tsx
Attempted import error: 'Button' is not exported from '@kaitok/mono-ui' (imported as 'Button').
```

But it renders.

![Screenshot 2023-06-14 at 14 57 01](https://github.com/vercel/next.js/assets/21013447/e1a875f2-f7e6-4c8d-b739-1757ad8f53d9)

**Reply by icyJoseph** [reply, marked as answer]
Alright, so yeah, it was that, you gotta make react and react-dom peerDependencies in your package.json, and then use the peerDepsExternal package.

```json
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "18.2.0"
  },
```

```js
import peerDepsExternal from "rollup-plugin-peer-deps-external";

// later on


  plugins: [
    peerDepsExternal(),
    resolve(),
    commonjs(),
    typescript({ tsconfig: "./tsconfig.json" }),
    postcss(),
  ],
```

The issue I have now is that, Node fails to import, the CJS files, it is such as mess with ESM and CJS, but yeah, it fails to render server side, but the client side finds the ESM modules...

**Reply by kaitok** [reply]
@icyJoseph Thank you! It worked. 
I see, so I had to specify react and react-dom in the peerDependencies to get the same version for library users.
